### DISCUSSION TITLE
[질문] 이 프로젝트에서 모듈마다 갖고 있는 특성과 그 역할은 무엇인가요? (사진 첨부)

### DISCUSSION THREAD

**Discussion by wonseok97** [TEAM-DAME/Dame-Android#39] [2. Q&A] [answered]
<img width="152" alt="image" src="https://user-images.githubusercontent.com/118143620/201555648-1b7a5db3-a72b-470f-8e67-feaf03c2c0bc.png">

**Comment by onseok** [comment, marked as answer]
모듈화 전략의 장점은 각각의 모듈마다 독립적으로 빌드되고 테스트가 가능하다는 점입니다.
또한 모듈들이 가능하다면 서로가 서로를 알지 못하기(독립적) 때문에, 한 모듈의 변화가 다른 모듈들에게 미치는 영향이 없거나 미미합니다. 이것을 `Low coupling`이라고 합니다. 또한 모듈은 하나의 시스템으로 동작되는 코드의 집합으로만 구성되어야 합니다. 예를 들어 `navigator` 모듈은 앱 내의 navigation과 관련된 인터페이스를 정의하는 코드로만 이루어져 있습니다. 이것을 `High cohesion`이라고 합니다. 

`app`모듈은 앱이 정상적으로 작동하도록 필요한 모든 것들을 가져옵니다. 그리고 이것은 `UI scaffolding`과 `navigation`을 담당합니다. 또한 `hilt module`등을 담당합니다. 

`buildSrc`모듈은 `Kotlin DSL`과 함께 프로젝트의 `Dependency`를 관리하기 위해 필요한 모듈입니다.

`core-data`모듈은 여러 모듈들에서 필요한 `data` 레이어 계층을 담당하는 보조적인 코드들을 포함합니다. 

`core-ui`모듈은 여러 모듈들에서 필요한 `presentation` 레이어 계층을 담당하는 보조적인 코드들을 포함합니다. 

`domain` 모듈은 의존성을 가지고 있지 않은 모듈(계층)이며, 순수 코틀린/자바 코드로만 이루어져 있습니다. 비즈니스 로직에 필요한 `data model`과 `UseCase`가 포함되어 있는 계층입니다. 

`data` 모듈은 `domain` 모듈(계층)에 의존성을 가지고 있으며, 말 그대로 data를 관리하는 모듈입니다. Repository 구현체, Cache, Room DB, Dao, API 통신과 그 결과로 가져오는 data entity, data 계층의 데이터(받아오는 데이터 형태)를 domain 계층의 데이터(실제로 사용되는 데이터 형태)로 변환해주는 `mapper` 클래스 등을 포함합니다.

`feature` 모듈은 클린 아키텍처 개념에서 각각 `presentation` 계층을 담당한다고 보면 됩니다. `data`와 `domain`계층에 의존성을 가지고 있기 때문에, 두 계층의 변화에 모두 영향을 받습니다. 예를 들어 `domain` 계층에서 선언된 `UseCase`는 `presentation`계층의 `ViewModel`에서 직접 사용하기 때문에 간접적인 것이 아닌 직접적으로 영향을 끼치게 됩니다. 현재 `다메다메` 프로젝트에서도 `hilt`를 사용하여 `usecases`를 `feature` 모듈의 `viewmodel` 클래스에 주입합니다.

마지막으로 `shared` 모듈은 프로젝트 공통으로 사용되는 리소스들을 관리합니다.
